### PR TITLE
Feature/init

### PR DESCRIPTION
**Pull request by chemoish**

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    var sourceUrl = document.getElementsByName('source_url')[0];
+    sourceUrl.value = localStorage.getItem('source_url');
+
+    var licenseUrl = document.getElementsByName('license_url')[0];
+    licenseUrl.value = localStorage.getItem('license_url');
+  </script>
+
+  <script>
+    function onBlur(event) {
```
Comment (m14t) [inline index.html:83]: NBD: If you moved this to the `<script>` tag above, could you just add `sourceUrl.onblur = onBlur` and not need the html tag?  It would keep all the local storage stuff together.

File: index.html [hunk 2]
```diff
@@ -0,0 +1,119 @@
+<!doctype html>
```
Comment (m14t) [inline index.html:1]: NBD: If possibly, you should publish this to the (`gh-pages`](https://help.github.com/articles/creating-project-pages-manually/) branch. This would let people play with it without having to clone anything.

http://rafeca.com/2012/01/17/automate-your-release-flow/ (step 5) has a neat way to keep the histories of the two branches related to each other.

File: src/fairplay.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { stringToArray } from './util';
+
+export default function concatInitDataIdAndCertificate(initData, id, certificate) {
+  if (typeof id === 'string') {
```
Comment (m14t) [inline src/fairplay.js:4]: Is there a reason that `id` is the only thing that we check if its a `string`?

File: src/fairplay.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    id = stringToArray(id);
```
Comment (m14t) [inline src/fairplay.js:5]: In general you don't want to reassign params?

http://eslint.org/docs/rules/no-param-reassign

File: src/fairplay.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  const size = initData.byteLength + 4 + id.byteLength + 4 + certificate.byteLength;
+  let offset = 0;
+
+  const buffer = new ArrayBuffer(size);
+
+  const dataView = new DataView(buffer);
+
+  const initDataArray = new Uint8Array(buffer, offset, initData.byteLength);
+  initDataArray.set(initData);
```
Comment (m14t) [inline src/fairplay.js:23]: Is this `initDataArray` needed/used?

File: src/util.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
+export function arrayToString(array) {
+  return String.fromCharCode.apply(null, new Uint16Array(array.buffer));
+}
+
+export function base64DecodeUint8Array(input) {
+  const raw = window.atob(input);
```
Comment (m14t) [inline src/util.js]: just `atob` ?

File: src/videojs-fairplay.js [hunk 7]
```diff
@@ -0,0 +1,239 @@
+/* global videojs, WebKitMediaKeys */
+
+import * as util from './util';
```
Comment (m14t) [inline src/videojs-fairplay.js]: Just `import { arrayToString, getHostnameFromURI } from './util';`

(only 60 chars)?

File: src/videojs-fairplay.js [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    if (!keySession) {
+      throw new Error('Could not create key session');
+    }
+
+    return keySession;
+  }
+
+  fetchCertificate({ callback }) {
+    const { certificateUrl } = this.protection_;
+
```
Comment (m14t) [inline src/videojs-fairplay.js:71]: No newline between `const`s?

File: src/videojs-fairplay.js [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+      });
+    }, false);
+
+    request.open('GET', certificateUrl, true);
+    request.send();
+  }
+
+  fetchLicense({ target, message }) {
+    const { licenseUrl } = this.protection_;
+
```
Comment (m14t) [inline src/videojs-fairplay.js:91]: Same?

File: src/videojs-fairplay.js [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+  fetchCertificate({ callback }) {
+    const { certificateUrl } = this.protection_;
+
+    const request = new XMLHttpRequest();
+
+    request.responseType = 'arraybuffer';
+
+    request.addEventListener('error', this.onCertificateError, false);
+    request.addEventListener('load', (event) => {
+      this.onCertificateLoad(event, {
```
Comment (m14t) [inline src/videojs-fairplay.js:78]: Why is the 2nd param here an object with a `callback` inside of it?  It seems that we could just pass the callback in directly.  I didn't see any places that called `onCertificateLoad` with anything else in that 2nd option.

File: src/videojs-fairplay.js [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+        callback,
+      });
+    }, false);
+
+    request.open('GET', certificateUrl, true);
+    request.send();
+  }
+
+  fetchLicense({ target, message }) {
```
Comment (m14t) [inline src/videojs-fairplay.js:87]: Is there a reason to make `fetchLicense` take a single object with two properties, instead of just 2 parameters?

File: src/videojs-fairplay.js [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+  onVideoError() {
+    this.player_.error({
+      code: 0,
+      message: 'A video playback error occurred.',
+    });
+  }
+
+  onVideoWebkitNeedKey(event) {
+    const { keySystem } = this.protection_;
+
```
Comment (m14t) [inline src/videojs-fairplay.js:205]: So many new lines.

File: src/videojs-fairplay.js [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    return this.fetchCertificate({
+      callback: () => {
+        this.tech_.src(src);
+      },
+    });
+  }
+
+  static setLogToBrowserConsole(value = false) {
```
Comment (m14t) [inline src/videojs-fairplay.js]: `static` methods at the top?

File: .eslintrc.yml [hunk 14]
```diff
@@ -0,0 +1,20 @@
+root: true
```
Comment (m14t) [inline .eslintrc.yml:1]: Should we setup a circle-ci job to run the linter on push?

File: index.html (hunk 1 above)
Comment (chemoish) [inline index.html:83]: I tried this and it was less explicit. Most of this code is unrelated and I would ideally like out of sight out of mind; however, I consolidated the script tags.

File: src/fairplay.js (hunk 3 above)
Comment (chemoish) [inline src/fairplay.js:4]: `id` is `contentId` which is based off of `initData`, in the previous function it converts the `event.initData`, a `Uint8Array`, into a string, into a anchor tag, to pull off the `hostname`, which leaves you at a string.

this function needs a bunch of `Uint8Array`'s and `contentId`, or `id` as seen above, is the only one that can be a string—needs to be converted back into a `Uint8Array`.

as to when it would never be a string? not sure,

File: src/fairplay.js (hunk 5 above)
Comment (chemoish) [inline src/fairplay.js:23]: I thought the same thing, but I left it in, going to test it without it.

(They end up being the same, so I am going to opt to use `intDataArray` to match the rest of the buffer creation)

File: src/fairplay.js (hunk 4 above)
Comment (chemoish) [inline src/fairplay.js:5]: In general yes, though I wanted to follow existing format to be as close to original HTML5 implementation as possible.

File: src/util.js (hunk 6 above)
Comment (chemoish) [inline src/util.js]: 👍

File: src/videojs-fairplay.js (hunk 7 above)
Comment (chemoish) [inline src/videojs-fairplay.js]: Was originally much bigger, will update.

File: src/videojs-fairplay.js (hunk 10 above)
Comment (chemoish) [inline src/videojs-fairplay.js:78]: It originally had a bunch of other things in it that was refactored out. Will update.

File: src/videojs-fairplay.js (hunk 13 above)
Comment (chemoish) [inline src/videojs-fairplay.js]: I think this is only a requirement of `eslint-plugin-react`? Will move it up anyways.

File: .eslintrc.yml (hunk 14 above)
Comment (chemoish) [inline .eslintrc.yml:1]: Ya, eventually.
